### PR TITLE
added clang-tidy CI job and fixed warnings / added CMake option "DISABLE_CPP03_SYNTAX_CHECK" / CMake cleanups

### PR DESCRIPTION
**Pull request by firewave**
The only interesting suppress warning is currently `clang-analyzer-core.NullDereference` since it depends on very complex value flow, might be a false positive and I didn't have the time to review it.

### CODE REVIEW

File: simplecpp.cpp [hunk 1]
```diff
@@ -221,15 +221,17 @@ simplecpp::TokenList::TokenList(std::istream &istr, std::vector<std::string> &fi
     readfile(istr,filename,outputList);
 }
 
-simplecpp::TokenList::TokenList(const TokenList &other) : frontToken(nullptr), backToken(nullptr), files(other.files)
+simplecpp::TokenList::TokenList(const TokenList &other)
```
Comment (danmar) [inline simplecpp.cpp:224]: I prefer to copy simplecpp in a separate PR.

**Comment by firewave** [comment]
I disabled `performance-inefficient-string-concatenation` since that warning appears to have false positives - see https://github.com/llvm/llvm-project/issues/54526. I was always curious about this warning but never had the time to properly look into it.

File: simplecpp.cpp [hunk 2]
```diff
@@ -2796,7 +2796,7 @@ static std::string getFileName(const std::map<std::string, simplecpp::TokenList
         return (filedata.find(header) != filedata.end()) ? simplecpp::simplifyPath(header) : "";
     }
 
-    const std::string relativeFilename = getRelativeFileName(sourcefile, header);
```
Comment (danmar) [inline simplecpp.cpp:2799]: I just wonder why you remove this const.

**Review by danmar — COMMENTED** [review]
just a small question

File: simplecpp.cpp (hunk 2 above)
Comment (firewave) [reply simplecpp.cpp:2799]: That's the `performance-no-automatic-move` check. If a variable is returned and it is declared `const` it cannot be moved but is copied instead.

Comment (danmar) [reply simplecpp.cpp:2799]: so the drawback is less safety. is a const reference a better solution perhaps?

Comment (firewave) [reply simplecpp.cpp]: I will drop this for now.

**Comment by firewave** [comment]
This is ready for review again. I also updated and modernized the clang-tidy job.

File: simplecpp.cpp [hunk 3]
```diff
@@ -1842,7 +1842,7 @@ namespace simplecpp {
                 return recursiveExpandToken(output, temp, loc, tok2, macros, expandedmacros2, parametertokens);
             }
 
-            else if (tok->str() == DEFINED) {
```
Comment (danmar) [inline simplecpp.cpp:1845]: please undo all changes in simplecpp.. these should be made in upstream and then they can be copied to cppcheck.

Comment (firewave) [reply simplecpp.cpp:1845]: We are currently in the simplecpp repo. 🙄

Comment (danmar) [reply simplecpp.cpp:1845]: oops!! :-)

**Pull request merged**